libraries/Microsoft.Bot.Builder.Ai/LocaleConverter.cs

```diff
@@ -0,0 +1,191 @@
+        /// Supporting English, French, Deutsche and Chinese Locales.
+        /// </summary>
+        private void initLocales()
+        {
+            DateAndTimeLocaleFormat yearMonthDay = new DateAndTimeLocaleFormat
+            {
+                TimeFormat = "{0:hh:mm tt}",
+                DateFormat = "{0:yyyy-MM-dd}"
+            };
+            DateAndTimeLocaleFormat dayMonthYear = new DateAndTimeLocaleFormat
+            {
+                TimeFormat = "{0:hh:mm tt}",
+                DateFormat = "{0:dd/MM/yyyy}"
+            };
+            DateAndTimeLocaleFormat monthDayYEar = new DateAndTimeLocaleFormat
+            {
+                TimeFormat = "{0:hh:mm tt}",
+                DateFormat = "{0:MM/dd/yyyy}"
+            };
+            foreach (string locale in new string[] { "en-za", "en-ie", "en-gb", "en-ca", "fr-ca", "zh-cn", "zh-sg", "zh-hk", "zh-mo", "zh-tw" })
+            {
+                mapLocaleToFunction[locale] = yearMonthDay;
+            }
+            foreach (string locale in new string[] { "en-au", "fr-be", "fr-ch", "fr-fr", "fr-lu", "fr-mc", "de-at", "de-ch", "de-de", "de-lu", "de-li" })
+            {
+                mapLocaleToFunction[locale] = dayMonthYear;
+            }
+            mapLocaleToFunction["en-us"] = monthDayYEar;
+        }
+        
+        /// <summary>
+        /// Check if a specific locale is available
+        /// </summary>
+        /// <param name="locale"></param>
+        /// <returns></returns>
+        public bool IsLocaleAvailable(string locale)
+        {
+            return mapLocaleToFunction.ContainsKey(locale);
+        }
+        
+        /// <summary>
+        /// Extract date and time from a sentence using Microsoft Recognizer
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="fromLocale">Source Locale</param>
+        /// <returns></returns>
+        private List<TextAndDateTime> extractDate(string message, string fromLocale)
+        {
+            List<TextAndDateTime> fndDates = new List<TextAndDateTime>();
+            var culture = Culture.English; 
+            if (fromLocale.StartsWith("fr"))
+            {
+                culture = Culture.French;
+            }
+            else if (fromLocale.StartsWith("de"))
+            {
+                culture = Culture.German;
+            }
+            else if (fromLocale.StartsWith("pt"))
+            {
+                culture = Culture.Portuguese;
+            }
+            else if (fromLocale.StartsWith("zh"))
+            {
+                culture = Culture.Chinese;
+            }
+            else if (fromLocale.StartsWith("es"))
+            {
+                culture = Culture.Spanish;
+            }
+            else if(!fromLocale.StartsWith("en"))
+            {
+                throw (new ArgumentException("Unsupported From Locale"));
+            }
+            var model = DateTimeRecognizer.GetInstance().GetDateTimeModel(culture);
```

Comment:
My understanding is that this call is *very* expensive, as it will construct all of the regular expressions. 

Should we consider caching this by culture?

Reply:
Caching by culture seems to be a great idea (done).